### PR TITLE
Enable nullable reference types for Hosting

### LINKED ISSUE #1760: Hosting Nullable reference types support
It would be better to support nullable reference types for all libraries

### PR DESCRIPTION
**Pull request by WeihanLi**
Fixes #1760 

- Enable nullable reference types for Hosting
- Renaming `commandline` => `commandLine` to fix typo
- Remove unnecessary `ToList` for configuration extension

### CODE REVIEW

File: src/System.CommandLine.Hosting/Directory.Build.props [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <UseAlphaVersion>true</UseAlphaVersion>
+    <Nullable>enable</Nullable>
```
Comment (jonsequitur) [inline src/System.CommandLine.Hosting/Directory.Build.props:5]: Please make this change in the `.csproj` file for consistency with the other projects.

File: src/System.CommandLine.Hosting/HostingExtensions.cs [hunk 2]
```diff
@@ -107,7 +107,7 @@ public static IHostBuilder UseCommandHandler(this IHostBuilder builder, Type com
                 && invocation.ParseResult.CommandResult.Command is Command command
                 && command.GetType() == commandType)
             {
-                invocation.BindingContext.AddService(handlerType, c => c.GetService<IHost>().Services.GetService(handlerType));
+                invocation.BindingContext.AddService(handlerType, c => c.GetRequiredService<IHost>().Services.GetService(handlerType));
```
Comment (jonsequitur) [inline src/System.CommandLine.Hosting/HostingExtensions.cs:110]: Why this change? I would think the `IHost` is a required service here.

File: src/System.CommandLine.Hosting/HostingExtensions.cs [hunk 3]
```diff
@@ -141,11 +141,14 @@ public static InvocationContext GetInvocationContext(this HostBuilderContext con
...
-        public static IHost GetHost(this InvocationContext invocationContext)
+        public static IHost? GetHost(this InvocationContext invocationContext)
         {
             _ = invocationContext ?? throw new ArgumentNullException(paramName: nameof(invocationContext));
             var hostModelBinder = new ModelBinder<IHost>();
-            return (IHost)hostModelBinder.CreateInstance(invocationContext.BindingContext);
+            var instance = hostModelBinder.CreateInstance(invocationContext.BindingContext);
+            if (instance is IHost host)
+              return host;            
+            return null;
```
Comment (jonsequitur) [inline src/System.CommandLine.Hosting/HostingExtensions.cs:151]: I would think we should  not allow null here.

File: src/System.CommandLine.Hosting/InvocationLifetime.cs [hunk 4]
```diff
@@ -17,7 +17,7 @@
 
 namespace System.CommandLine.Hosting
 {
-    public class InvocationLifetime : IHostLifetime
+    public sealed class InvocationLifetime : IHostLifetime
```
Comment (jonsequitur) [inline src/System.CommandLine.Hosting/InvocationLifetime.cs:20]: What is  the reason for this change?

File: src/System.CommandLine.Hosting/InvocationLifetimeOptions.cs [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 namespace System.CommandLine.Hosting
 {
-    public class InvocationLifetimeOptions : ConsoleLifetimeOptions
+    public sealed class InvocationLifetimeOptions : ConsoleLifetimeOptions
```
Comment (jonsequitur) [inline src/System.CommandLine.Hosting/InvocationLifetimeOptions.cs:5]: What is the reason for this change?

File: src/System.CommandLine.Hosting/HostingExtensions.cs (hunk 2 above)
Comment (WeihanLi) [reply src/System.CommandLine.Hosting/HostingExtensions.cs:110]: `GetRequiredService` would return service and ensure the service is not null, `GetService` would return `null` when no service found, think we expect `IHost` exists, so maybe better to use `GetRequiredService` 
<https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.serviceproviderserviceextensions.getrequiredservice?view=dotnet-plat-ext-6.0>

File: src/System.CommandLine.Hosting/InvocationLifetime.cs (hunk 4 above)
Comment (WeihanLi) [reply src/System.CommandLine.Hosting/InvocationLifetime.cs:20]: From this issue https://github.com/dotnet/runtime/issues/49944, sealed types would get performance gains and there're no virtual methods to override so think it could be a sealed class, same for the `InvocationLifetimeOptions`

File: src/System.CommandLine.Hosting/HostingExtensions.cs (hunk 3 above)
Comment (WeihanLi) [reply src/System.CommandLine.Hosting/HostingExtensions.cs:151]: I think so before, from test code <https://github.com/dotnet/command-line-api/blob/209b724a3c843253d3071e8348c353b297b0b8b5/src/System.CommandLine.Hosting.Tests/HostingTests.cs#L401-L403>
it may be null, if this should not be `null`, guess we should throw an `InvalidOperationException`?

File: src/System.CommandLine.Hosting/InvocationLifetimeOptions.cs (hunk 5 above)
Comment (WeihanLi) [reply src/System.CommandLine.Hosting/InvocationLifetimeOptions.cs]: Reverted

File: src/System.CommandLine.Hosting/InvocationLifetime.cs (hunk 4 above)
Comment (WeihanLi) [reply src/System.CommandLine.Hosting/InvocationLifetime.cs]: Reverted

File: src/System.CommandLine.Hosting/Directory.Build.props (hunk 1 above)
Comment (WeihanLi) [reply src/System.CommandLine.Hosting/Directory.Build.props]: Updated

File: src/System.CommandLine.Hosting/InvocationLifetime.cs (hunk 4 above)
Comment (jonsequitur) [reply src/System.CommandLine.Hosting/InvocationLifetime.cs]: This is good performance advice but generally something I'd avoid applying universally without a benchmark indicating that these specific types will benefit from being `sealed`.

File: src/System.CommandLine.Hosting/HostingExtensions.cs (hunk 2 above)
Comment (WeihanLi) [reply src/System.CommandLine.Hosting/HostingExtensions.cs:110]: Think it should not be registered when `IHost` or the handler is `null`, maybe I could add a null check for that

File: src/System.CommandLine.Hosting/HostingExtensions.cs (hunk 3 above)
Comment (Keboo) [inline src/System.CommandLine.Hosting/HostingExtensions.cs:151]: More succinct syntax could be (assuming we keep the null return):
```suggestion
            return hostModelBinder.CreateInstance(invocationContext.BindingContext) as IHost;
```
